File: Backend/src/main/java/backend/osiris/entity/promotion/Promotion.java
package backend.osiris.entity.promotion;

import backend.osiris.entity.BaseEntity;
import backend.osiris.entity.product.Product;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.time.Instant;
import java.util.HashSet;
import java.util.Set;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
@Entity
@Table(name = "promotion")
public class Promotion extends BaseEntity {
    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "start_date", nullable = false)
    private Instant startDate;

    @Column(name = "end_date", nullable = false)
    private Instant endDate;

    @Column(name = "percent", nullable = false)
    private Integer percent;

    @Column(name = "status", nullable = false, columnDefinition = "TINYINT")
    private Integer status;

    @ManyToMany(cascade = {CascadeType.MERGE})
    @JoinTable(
            name = "promotion_product",
            joinColumns = @JoinColumn(name = "promotion_id", nullable = false),
            inverseJoinColumns = @JoinColumn(name = "product_id", nullable = false)
    )
    private Set<Product> products = new HashSet<>();
}

File: Frontend/src/models/Promotion.ts
import BaseResponse from "./BaseResponse";
import { ProductResponse } from "./Product";

export interface PromotionResponse extends BaseResponse {
    name: string;
    startDate: string;
    endDate: string;
    percent: number;
    status: number;
    products: ProductResponse[];
}

export interface PromotionRequest {
    name: string;
    startDate: string;
    endDate: string;
    percent: number;
    status: number;
    productIds: number[];
    categoryIds: number[];
}

export interface PromotionCheckingResponse {
    promotionable: boolean;
}
